### ISSUE TITLE
Premature flush of limbo during promote

### ISSUE THREAD

**Issue by funbringer** [tarantool/tarantool#9376]
**Bug description**

While node is being promoted, `txn_limbo_write_promote` writes a `RAFT_PROMOTE` request to its wal and immediately proceeds to flush the limbo in `txn_limbo_read_promote`:

https://github.com/tarantool/tarantool/blob/4b88d5fd9833b1c236572e7c557646909ec415eb/src/box/txn_limbo.c#L576-L579

https://github.com/tarantool/tarantool/blob/4b88d5fd9833b1c236572e7c557646909ec415eb/src/box/txn_limbo.c#L1196-L1198

https://github.com/tarantool/tarantool/blob/4b88d5fd9833b1c236572e7c557646909ec415eb/src/box/txn_limbo.c#L591-L593

However, according to the raft paper (3.6.2),

> As described in Section 3.5, a leader knows that an entry from its current term is committed once that entry is stored on a majority of the servers. If a leader crashes before committing an entry, future leaders will attempt to finish replicating the entry. However, **a leader cannot immediately conclude that an entry from a previous term is committed once it is stored on a majority of servers.**
> ...
> To eliminate problems like the one in Figure 3.7, Raft never commits log entries from previous terms by counting replicas. Only log entries from the leader’s current term are committed by counting replicas; once an entry from the current term has been committed in this way, then all prior entries are committed indirectly because of the Log Matching Property. There are some situations where a leader could safely conclude that an older log entry is committed (for example, if that entry is stored on every server), but Raft takes a more conservative approach for simplicity.

Until the newly elected leader has **committed** a new raft entry in the current term, it is still possible to rewrite the history using an entry with a newer term. If we consider that rewinding a limbo flush is not possible, this means that the current behavior is not correct. Furthermore, this causes problems in production when the current leader writes `RAFT_PROMOTE` but fails to deliver it to other nodes on time, causing another "split brain" scenario.

TL;DR: We should only apply the side effects of promote after a quorum has been reached for a new entry with the current term (e.g. the `RAFT_PROMOTE` itself).

**Comment by sergepetrenko** [comment]
See also: #8095
